src/kernel/disconnected_transactions.h

```diff
@@ -0,0 +1,124 @@
+private:
+    uint64_t cachedInnerUsage = 0;
+    const size_t m_max_mem_usage;
+    std::list<CTransactionRef> queuedTx;
+    using List = decltype(queuedTx);
+    std::unordered_map<uint256, List::iterator, SaltedTxidHasher> iters_by_txid;
+
+public:
+    DisconnectedBlockTransactions(size_t max_mem_usage) : m_max_mem_usage{max_mem_usage} {}
+
+    // It's almost certainly a logic bug if we don't clear out queuedTx before
+    // destruction, as we add to it while disconnecting blocks, and then we
+    // need to re-process remaining transactions to ensure mempool consistency.
+    // For now, assert() that we've emptied out this object on destruction.
+    // This assert() can always be removed if the reorg-processing code were
+    // to be refactored such that this assumption is no longer true (for
+    // instance if there was some other way we cleaned up the mempool after a
+    // reorg, besides draining this object).
+    ~DisconnectedBlockTransactions() {
+        assert(queuedTx.empty());
+        assert(iters_by_txid.empty());
+        assert(cachedInnerUsage == 0);
+    }
+
+    size_t DynamicMemoryUsage() const {
+        return cachedInnerUsage + memusage::DynamicUsage(iters_by_txid) + memusage::DynamicUsage(queuedTx);
+    }
+
+    // Add transactions from the block, in reverse order. We assume that callers will never give us
+    // multiple transactions with the same txid, otherwise things can go very wrong in
+    // removeForBlock due to queuedTx containing an item without a corresponding entry in iters_by_txid.
+    // Returns vector of transactions that were evicted for size-limiting.
+    [[nodiscard]] std::vector<CTransactionRef> AddTransactionsFromBlock(const std::vector<CTransactionRef>& vtx)
+    {
+        std::vector<CTransactionRef> evicted;
+        // Blocks are disconnected in descending order by block height. Within each block, add
+        // transactions in reverse order so that transactions with dependencies on other
+        // transactions (if any) are at the beginning. If this data structure grows too large, we
+        // will trim transactions from the front, and should avoid removing transactions that other
+        // ones depend on.
+        iters_by_txid.reserve(iters_by_txid.size() + vtx.size());
+        for (auto block_it = vtx.rbegin(); block_it != vtx.rend(); ++block_it) {
+            auto it = queuedTx.insert(queuedTx.end(), *block_it);
+            iters_by_txid.emplace((*block_it)->GetHash(), it);
+            cachedInnerUsage += RecursiveDynamicUsage(*block_it);
+        }
+
+        // Trim the earliest-added entries until we are within memory bounds.
+        while (!queuedTx.empty() && DynamicMemoryUsage() > m_max_mem_usage) {
+            evicted.emplace_back(queuedTx.front());
+            cachedInnerUsage -= RecursiveDynamicUsage(queuedTx.front());
+            iters_by_txid.erase(queuedTx.front()->GetHash());
+            queuedTx.pop_front();
+        }
+        return evicted;
```

Comment:
For readability, would you consider moving this to its own private function?

<details>
<summary>git diff</summary>

```diff
diff --git a/src/kernel/disconnected_transactions.h b/src/kernel/disconnected_transactions.h
index 6e12059874..d1a8b5e989 100644
--- a/src/kernel/disconnected_transactions.h
+++ b/src/kernel/disconnected_transactions.h
@@ -37,6 +37,20 @@ private:
     using List = decltype(queuedTx);
     std::unordered_map<uint256, List::iterator, SaltedTxidHasher> iters_by_txid;
 
+    // Trim and return the earliest-added entries until we are within memory bounds.
+    std::vector<CTransactionRef> EvictTransactionsIfNeeded()
+    {
+        std::vector<CTransactionRef> evicted;
+        while (!queuedTx.empty() && DynamicMemoryUsage() > m_max_mem_usage) {
+            evicted.emplace_back(queuedTx.front());
+            cachedInnerUsage -= RecursiveDynamicUsage(queuedTx.front());
+            iters_by_txid.erase(queuedTx.front()->GetHash());
+            queuedTx.pop_front();
+        }
+
+        return evicted;
+    }
+
 public:
     DisconnectedBlockTransactions(size_t max_mem_usage) : m_max_mem_usage{max_mem_usage} {}
 
@@ -64,7 +78,6 @@ public:
     // Returns vector of transactions that were evicted for size-limiting.
     [[nodiscard]] std::vector<CTransactionRef> AddTransactionsFromBlock(const std::vector<CTransactionRef>& vtx)
     {
-        std::vector<CTransactionRef> evicted;
         // Blocks are disconnected in descending order by block height. Within each block, add
         // transactions in reverse order so that transactions with dependencies on other
         // transactions (if any) are at the beginning. If this data structure grows too large, we
@@ -77,13 +90,7 @@ public:
             cachedInnerUsage += RecursiveDynamicUsage(*block_it);
         }
 
-        // Trim the earliest-added entries until we are within memory bounds.
-        while (!queuedTx.empty() && DynamicMemoryUsage() > m_max_mem_usage) {
-            evicted.emplace_back(queuedTx.front());
-            cachedInnerUsage -= RecursiveDynamicUsage(queuedTx.front());
-            iters_by_txid.erase(queuedTx.front()->GetHash());
-            queuedTx.pop_front();
-        }
+        const auto evicted{EvictTransactionsIfNeeded()};
         return evicted;
     }
 
```
</details>